### PR TITLE
demo: Add Makefile to simplify steps and self document

### PR DESCRIPTION
**Pull request by simple10**
Signed-off-by: Joe Johnston <joe@simple10.com>

This will hopefully make it a lot easier for people to get quickly up and running with flynn.

It also fixes https://github.com/flynn/flynn/issues/975

### CODE REVIEW

File: demo/Makefile [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+# Providers: virtualbox, vmware_fusion, aws
+PROVIDER := virtualbox
+
+
+.PHONY: install init bootstrap up down halt update ssh reload reset token destroy dashboard ssh-cmd
+
+
+# Install flynn cli on host
+install:
```
Comment (titanous) [inline demo/Makefile]: I'd call this target `cli`

File: demo/Makefile [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	@$(MAKE) CMD='sudo stop flynn-host' ssh-cmd
+
+# Halt the VM
+halt:
+	vagrant halt
+
+# Update flynn in the VM
+# VM must already be up and running
+update:
+	@$(MAKE) CMD='sudo bash < <(curl -fsSL https://dl.flynn.io/install-flynn) && sudo flynn-host update' ssh-cmd
```
Comment (titanous) [inline demo/Makefile]: The second command is unnecessary, currently the updater is incomplete so installing the latest version requires restarting from scratch.

File: demo/Vagrantfile [hunk 3]
```diff
@@ -22,15 +24,17 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
 
-  config.vm.provision "shell", privileged: false, inline: <<SCRIPT
-    ## Uncomment these lines to update flynn before starting
-    # sudo apt-get update -qq
-    # sudo apt-get install -y flynn-host
-    # sudo flynn-host download /etc/flynn/version.json
-
-    sudo start flynn-host
+  config.vm.provider "virtualbox" do |v|
+    v.gui = true
```
Comment (titanous) [inline demo/Vagrantfile]: Why this flag?

File: demo/Makefile (hunk 2 above)
Comment (simple10) [inline demo/Makefile]: flynn-host does not currently auto start after halting the VM and restarting it. Thus the addition of `start flynn-host` in `make up`.

Should I just add a comment that this should be removed once `flynn-host` is added to the init system?

File: demo/Vagrantfile (hunk 3 above)
Comment (simple10) [inline demo/Vagrantfile]: Glad you asked. ;-)

I've worked with a lot of teams on-boarding them onto Vagrant. One of the really annoying things about Vagrant to someone who hasn't used it much (like a developer moving from Heroku to Flynn) is how it somewhat obfuscates the fact that `vagrant up` actually started a VM that is hogging a lot of resources. Or more precise, the fact that the dev has to run a `vagrant halt` to stop the VM from hogging resources. This is obvious to anyone who has Vagrant experience, but surprisingly easy to forget for non-Vagrant users.

Adding the gui flag causes VirtualBox and VMWare to show the normal KVM'ish interface and serves as a visual reminder that the VM is running.

Basically, I'm in favor of as little magic as possible when people are demoing things.

File: demo/Makefile (hunk 2 above)
Comment (simple10) [inline demo/Makefile]: I might have misunderstood your comment @titanous. Did you mean `make down` is unnecessary or the `start flynn-host` in `make up` is unnecessary?

Comment (titanous) [inline demo/Makefile]: This is about `make update`: I think it's probably less confusing to replace the command with `vagrant box update` for now until the in-place updater is finished.

Comment (simple10) [inline demo/Makefile]: Got it. Pushed.

File: demo/Makefile [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	@$(MAKE) up
+	@$(MAKE) bootstrap
+	flynn key add
+
+# Bootstrap the cluster
+# This should only be called once per cluster
+# Runs bootstrap in VM and cluster add on host
+bootstrap:
+	@$(MAKE) CMD='CLUSTER_DOMAIN="${CLUSTER_DOMAIN}" flynn-host bootstrap /etc/flynn/bootstrap-manifest.json 2>&1' ssh-cmd | tee ${BOOTSTRAP_LOG_FILE}
+	cat ${BOOTSTRAP_LOG_FILE} | awk '/flynn cluster add/' | awk 'BEGIN {RS=""}{gsub(/\r/,"",$$0); print $$0}' | sh -s
```
Comment (lmars) [inline demo/Makefile]: We prefer using `bash` rather than `sh`, and `grep` for matching, so I suggest:

```
grep '^flynn cluster add' ${BOOTSTRAP_LOG_FILE} | bash
```

It may also be worth exiting non-zero if bootstrapping fails, e.g.:

```
(grep '^flynn cluster add' bootstrap.log || echo exit 1) | bash
```

File: demo/Vagrantfile (hunk 3 above)
Comment (titanous) [inline demo/Vagrantfile]: Ah, I see where you are coming from. I agree that for new users having no indicator that it is running can lead to forgetting about it. The current downside is that the window shows a login terminal that is basically useless, so we need to do something about that. A text or graphical message instead of the login prompt that explains how to access the dashboard would be one solution to this.

Comment (simple10) [inline demo/Vagrantfile]: Yeah, help text output in the box terminal would be really cool.

How about I comment out the gui flag for now and open a separate issue for reinstating it once the flynn box can be updated with help text output?

I'm not sure right off hand how to customize the actual login prompt on Debian/Ubuntu.

Comment (titanous) [inline demo/Vagrantfile]: Sounds good.

File: demo/Makefile (hunk 4 above)
Comment (simple10) [inline demo/Makefile]: I'd recommend sticking with `sh` when possible since it's universal where `bash` is sometimes not installed (mostly in cases where it's been replaced with zsh and removed from the system or locked down). But `bash` is fine for the demo. I'll update.

File: demo/Vagrantfile (hunk 3 above)
Comment (lmars) [inline demo/Vagrantfile]: FYI the contents of `/etc/issue` are printed before the login prompt, see [`login(5)`](http://linux.die.net/man/5/issue)

File: demo/Makefile (hunk 4 above)
Comment (lmars) [inline demo/Makefile]: > I'd recommend sticking with sh when possible since it's universal where bash is sometimes not installed

Yes I agree when portability is an issue, but here we have control of what is installed in the VM so I much prefer using `bash` whenever possible (I would even consider going as far as setting the default `sh` to `bash` rather than `dash`, not right now though :smile:)

File: demo/Vagrantfile (hunk 3 above)
Comment (simple10) [inline demo/Vagrantfile]: Oh sweet @lmars. How do I rebuild and test the box image? Looks like you're using packer.

File: demo/Makefile (hunk 4 above)
Comment (simple10) [inline demo/Makefile]: In the case with the Makefile, it's actually running on the host and not in the VM. So portability is a concern. But I've updated it to bash anyway.

File: demo/Makefile [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	vagrant box update
+
+# SSH into VM
+ssh:
+	vagrant ssh
+
+# Stop, start, and reprovision VM
+# Run to fix configuration or provisioning problems
+reload:
+	vagrant reload --provision
```
Comment (titanous) [inline demo/Makefile]: This doesn't redo any provisioning, could call it `reboot` or change it to `destroy` + `init`.

File: demo/Makefile (hunk 4 above)
Comment (lmars) [inline demo/Makefile]: Ah yes of course, my mistake.

File: demo/Makefile (hunk 5 above)
Comment (simple10) [inline demo/Makefile]: `vagrant reload --provision` reruns the provisioners. It's equivalent to ...

```bash
vagrant halt
vagrant up --provision
```

I only included it because `vagrant reload` is mentioned in the docs. But I've always thought "reload" is a confusing term.

How about calling it "reset" and just keeping it in there for troubleshooting purposes. i.e. Whenever someone shows up on Github with an issue related to the demo, the first response can be "Well, did you try running `make reset` ?"  ;-)

Comment (simple10) [inline demo/Makefile]: Oh, and we should probably add the box update as part of the reset.

Comment (titanous) [inline demo/Makefile]: Sounds good. There are no provisioners that aren't commented out btw.

Comment (simple10) [inline demo/Makefile]: Yeah, strictly there for future proofing because everybody forgets to update the Makefile after changing Vagrantfiles.  ;-)

File: demo/Makefile [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	vagrant box update
+
+# SSH into VM
+ssh:
+	vagrant ssh
+
+# Destroy and rebuilt the VM
+reset:
+	@$(MAKE) destroy
+	@$(MAKE) init
```
Comment (titanous) [inline demo/Makefile]: I think we discussed adding an `make update` in here?

File: demo/Makefile [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+# Stop flynn
+down:
+	@$(MAKE) CMD='sudo stop flynn-host' ssh-cmd
+
+# Halt the VM
+halt:
+	vagrant halt
+
+# Update the VM image to the lastest flynn vagrant box
```
Comment (titanous) [inline demo/Makefile]: s/lastest/latest

File: demo/Makefile (hunk 6 above)
Comment (simple10) [inline demo/Makefile]: `make update` is currently doing a `vagrant box update`. Can you clarify your comment?

Comment (titanous) [inline demo/Makefile]: https://github.com/flynn/flynn/pull/1304#discussion_r26723070

> Oh, and we should probably add the box update as part of the reset.

**Comment by titanous** [comment]
Is there anything left on this? I think it needs a squash and rebase on master before we can merge.

File: demo/Makefile [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+# Get token for API and dashboard
+token:
+	@flynn -a dashboard env get LOGIN_TOKEN
+
+# Open dashboard in browser
+dashboard:
+	@echo Login token:
+	@$(MAKE) token
+	@sleep 1 # Allow user to see token output before browser opens
+	@open https://dashboard.${CLUSTER_DOMAIN}
```
Comment (titanous) [inline demo/Makefile]: This should be `http`, no `s` so that the CA certificate flow works.

File: demo/Makefile [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+# Providers: virtualbox, vmware_fusion, aws
+PROVIDER := virtualbox
+
+
+.PHONY: install-cli init bootstrap up down halt update ssh reset token destroy dashboard ssh-cmd
+
+
+# Install flynn cli on host
+install-cli:
+	L=/usr/local/bin/flynn && curl -sL -A "`uname -sp`" https://dl.flynn.io/cli | zcat >$L && chmod +x $L
```
Comment (titanous) [inline demo/Makefile]: This doesn't work for me:

```text
L=/usr/local/bin/flynn && curl -sL -A "`uname -sp`" https://dl.flynn.io/cli | zcat > && chmod +x
/bin/sh: -c: line 0: syntax error near unexpected token `&&'
/bin/sh: -c: line 0: `L=/usr/local/bin/flynn && curl -sL -A "`uname -sp`" https://dl.flynn.io/cli | zcat > && chmod +x '
make: *** [install-cli] Error 2
```

File: demo/Makefile [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+# Initialize a new cluster
+init:
```
Comment (titanous) [inline demo/Makefile]: `init` should be the first target in the file so that a bare `make` triggers it.

File: demo/Makefile [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+	@$(MAKE) bootstrap
+	flynn key add
+	@printf "\nFlynn is now running in the VM.\nOpen dashboard: make dashboard\n\n"
+
+# Bootstrap the cluster
+# This should only be called once per cluster
+# Runs bootstrap in VM and cluster add on host
+bootstrap:
+	@$(MAKE) CMD='CLUSTER_DOMAIN="${CLUSTER_DOMAIN}" flynn-host bootstrap /etc/flynn/bootstrap-manifest.json 2>&1' ssh-cmd | tee ${BOOTSTRAP_LOG_FILE}
+	@(grep '^flynn cluster add .*$$' ${BOOTSTRAP_LOG_FILE} || echo exit 1) | sed "s/$$(printf '\r')\$$//" | bash
```
Comment (titanous) [inline demo/Makefile]: I think the `cluster add` command should be a separate target. It currently fails when I already have a cluster named `default` configured, so I think we should call the cluster `virtualbox` and set the `--force --default` flags when adding.

File: demo/Makefile [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+	rm ${BOOTSTRAP_LOG_FILE}
+
+# Start the VM and flynn
+up:
+	CLUSTER_DOMAIN="${CLUSTER_DOMAIN}" vagrant up --provider ${PROVIDER}
```
Comment (titanous) [inline demo/Makefile]: This `vagrant up` doesn't work for me, as I need to set `SSH_AUTH_SOCK=''` to work around mitchellh/vagrant#4367. I think we should set that env var for all vagrant commands, there's no downside as we're using the insecure key, so the SSH agent is not needed.

File: demo/Makefile [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	@$(MAKE) CMD='sudo start flynn-host || echo flynn-host already running or unable to start' ssh-cmd
+
+# Stop flynn
+down:
+	@$(MAKE) CMD='sudo stop flynn-host' ssh-cmd
```
Comment (titanous) [inline demo/Makefile]: This `down` doesn't have symmetry with `up`. I'd drop it and rename `halt` to `down`.

**Comment by titanous** [comment]
I added a few more comments after testing this out locally.

File: demo/Makefile [hunk 14]
```diff
@@ -0,0 +1,80 @@
+
+# Domain to use when starting flynn
+CLUSTER_DOMAIN := demo.localflynn.com
+CLUSTER := default
+BOOTSTRAP_LOG_FILE := bootstrap.${CLUSTER}.log
```
Comment (titanous) [inline demo/Makefile]: I don't think we need the cluster name in the log file. Also, it should be gitignored.

File: demo/Makefile (hunk 11 above)
Comment (simple10) [inline demo/Makefile]: How about calling the cluster "demo-virtualbox"?

Comment (titanous) [inline demo/Makefile]: Fine with me.

Comment (simple10) [inline demo/Makefile]: How do I set the name of the cluster? It seems like the `flynn-host bootstrap` command should include the cluster name since it outputs the command that's fed into bash. But I don't see a way in the help output to change the cluster name.

We could grep for `default` in the output but this seems kludgy.

Comment (titanous) [inline demo/Makefile]: Best bet is to rewrite the `cluster add` command since we need to add the force/default flags anyway. That message is only read by people using the manual install script, all of the orchestration tools use the JSON output from bootstrap and build the command using data from the relevant steps.

File: demo/Makefile [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+# Destroy VM
+# Run this to delete the VM and free up disk space
+destroy:
+	vagrant destroy -f
+	flynn cluster remove ${CLUSTER}
+
+# Get token for API and dashboard
+token:
+	@flynn -a dashboard env get LOGIN_TOKEN
```
Comment (titanous) [inline demo/Makefile]: This should specify `-c ${CLUSTER}`.

File: demo/Makefile [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+
+.PHONY: init install-cli bootstrap up down halt update ssh reset token destroy dashboard ssh-cmd
+
+
+# Initialize a new cluster
+init:
+	@$(MAKE) up
+	@$(MAKE) bootstrap
+	flynn key add
```
Comment (titanous) [inline demo/Makefile]: Ditto on the cluster specification.

**Comment by simple10** [comment]
I think this is ready for merge.

File: demo/Makefile [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+# Open dashboard in browser
+dashboard:
+	@echo Login token:
+	@$(MAKE) token
+	@sleep 1 # Allow user to see token output before browser opens
+	@open http://dashboard.${CLUSTER_DOMAIN}
+
+# Run a command in the VM via SSH
+ssh-cmd:
+	vagrant ssh -c '${CMD}'
```
Comment (titanous) [inline demo/Makefile]: `SSH_AUTH_SOCK` needs to be specified here and in the `ssh` target. Might make sense to make a variable.

**Comment by simple10** [comment]
This should be ready to merge.

**Comment by titanous** [comment]
@simple10 We just shipped Windows support for the CLI, and the current Vagrant setup (without this Makefile) should work fine on Windows. The Makefile will probably work fine if you have installed `make` somehow on Windows, but this is not ideal. Any thoughts on what we can do to keep the Windows support we already have while retaining the improved experience?

**Comment by simple10** [comment]
@titanous  In the past with other Vagrant projects, I've passed in ENV flags to the Vagrantfile to enable/disable features. So by default, the provisioners would run everything unless the disable flags were present. The Makefile would then pass in the disable flags to take finer grain control of provisioning steps.

I suggest we use this approach for now since it requires minimal changes to code and docs. Then, if someone wants to write a bat script for Windows, they can submit another pull request.

Sound good?

**Comment by titanous** [comment]
Yeah, that's fine.

**Comment by simple10** [comment]
Pushed. Works fine with or without the Makefile now.

**Comment by titanous** [comment]
LGTM! Thanks for your patience!
